fiftyone/utils/youtubedl.py

```diff
@@ -0,0 +1,343 @@
+|
+"""
+from collections import defaultdict
+import logging
+import multiprocessing
+import multiprocessing.dummy
+import os
+
+import eta.core.utils as etau
+import eta.core.video as etav
+
+import fiftyone.core.utils as fou
+
+youtube_dl = fou.lazy_import(
+    "youtube_dl",
+    callback=lambda: fou.ensure_import(
+        "youtube_dl",
+        error_msg="youtube_dl not found, run `pip install youtube-dl`",
+    ),
+)
+
+
+logger = logging.getLogger(__name__)
+
+
+def download_from_youtube(
+    urls,
+    download_dir=None,
+    clip_segments=None,
+    max_videos=None,
+    num_workers=None,
+    ext=None,
+    buffer_seconds=5,
+):
+    """
+    Attempts to download a list of video urls from YouTube.
+    The `urls` argument either accepts:
+
+        * A list of YouTube video urls::
+
+            urls = ["https://www.youtube.com/watch?v=-0URMJE8_PB", ...]
+
+          When `urls` is a list, then the `download_dir` argument is required
+          and all videos will be downloaded into that directory
+
+        * A dictionary mapping the video urls to files on disk in which to store that video::
+
+            urls = {
+                "https://www.youtube.com/watch?v=-0URMJE8_PB": "/path/to/local/file1.ext",
+                ...
+            }
+
+    The corresponding `clip_segments` argument can then contain a list oftuples of ints
+    or floats used to download segment clips of each video. A value of `None`
+    can indicate either downloading the entire video, downloading from the
+    start or downloading to the end of the video::
+
+        clip_segments = [
+            (10, 25),
+            (11.1, 20.2),
+            None,
+            (None, 8.3),
+        ]
+
+    Args:
+        urls: either a list of video urls, or a dict
+            mapping these urls to locations on disk. If `urls` is a list, then
+            the `download_dir` argument is required
+        download_dir (None): the output directory used to store the downloaded
+            videos. This is only used if `urls` is a list
+        clip_segments (None): a list of int or float tuples indicating start
+            and end times in seconds for downloading segments of videos
+        max_videos (None): the maximum number of videos to download from the
+            given ``urls``. By default, all ``urls`` will be downloaded
+        num_workers (None): the number of processes to use when downloading
+            individual video. By default, ``multiprocessing.cpu_count()`` is
+            used
+        ext (None): the extension to use to store the downloaded videos or a
+            list of extensions corresponding to the given ``urls``. By default,
+            the default extension of each video is used 
+        buffer_seconds (5): the number of additional seconds before and after
+            each clip to download to ensure nearest keyframes are captured if
+            `clip_segments` is provided
+
+    Returns:
+        urls of successfully downloaded videos
```

Comment:
how about a dict mapping URLs to local paths, for convenience?

Reply:
It would also be more useful if the errors dict also mapped `urls -> error messages`. The input is basically a list of URLs, so the most likely actions that will be taken based on the output is to check each `url` to see what happened to it.

Reply:
A dict for URLs doesn't really work since if you are downloading clips, you may have multiple local paths for one URL. I like the idea of returning url/local path tuples though.

Reply:
A list of tuples wouldn't quite work either in the sense that if you provided a list of clip segments to download that includes multiple clips from the same video and only one of the clips was actually downloaded because of a `max_videos` parameter, you wouldn't be able to deduce from the return values which of the clips was actually downloaded.

How about instead returning dicts that map integer index into the provided lists of `urls` and `clip_segments`. Both for successful downloads and errors. This would allow the caller to figure out what was attempted and succeeded, what was attempted and failed, and what was not attempted at all.